### PR TITLE
MINI-3488 Add Support for latest admob sdk

### PR DESCRIPTION
**Pull request by munir-rakuten**
# Description
 Need to add the `admob` module to Host App to make it work, if the module is not added it will use the old sdk if added will use the latest one

## Links
MINI-3488

# Checklist
- [x] I have read the [contributing guidelines](../.github/CONTRIBUTING.md).
- [x] I have completed the [SDK Development Learning Path](../.github/CONTRIBUTING.md#SDK-Development-Learning-Path) (if submitting a feature PR)
- [ ] I wrote/updated tests for new/changed code
- [x] I removed all sensitive data before every commit, including API endpoints and keys
- [x] I ran `./gradlew check` without errors

### CODE REVIEW

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/MiniAppDisplayActivity.kt
Comment (khairul-alam-licon) [inline testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/MiniAppDisplayActivity.kt:177]: I haven't completely finished the testing when set `false` here.

**Review by khairul-alam-licon — COMMENTED** [review]
Looking good, I have left few suggestions as initial start.

File: testapp/build.gradle
Comment (khairul-alam-licon) [inline testapp/build.gradle:146]: Should we document how to use this latest admob in hostapp? 🤔

**Review by khairul-alam-licon — COMMENTED** [review]
Looks good to me. 
Also, we need to add documentation in USERGUIDE & changelog as well.

File: testapp/build.gradle
Comment (munir-rakuten) [reply testapp/build.gradle:146]: @khairul-alam-licon  not yet.. after @corycaywood 's approval i will add it in userguide and changelog, thanks

Comment (Climbatize) [reply testapp/build.gradle:146]: I don't agree @munir-rakuten, we should not merge this PR before it is documented / added in changelog. If you think this PR does not need documentation because it might not pass the peer validation, submit it as a draft PR ;)

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/ads/AdMobDisplayer.kt
Comment (Climbatize) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/ads/AdMobDisplayer.kt:160]: forcing unwrapping with `!!` should be changed in my opinion. I'm not very good with kotlin, but is this possible?

```suggestion
        interstitialAdMap[adUnitId]?.let { 
            if (interstitialAdMap.containsKey(adUnitId) && it.isLoaded) {
                val ad = it
                ad.adListener = object : AdListener() {
                    override fun onAdClosed() {
                        onClosed.invoke()
                        interstitialAdMap.remove(adUnitId)
                    }
                }
                ad.show()
            } else
                onFailed.invoke("Ad is not loaded yet")
        }
```

File: testapp/build.gradle
Comment (munir-rakuten) [reply testapp/build.gradle:146]: @Climbatize got your point 🙂, thanks

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/ads/AdMobDisplayer.kt
Comment (munir-rakuten) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/ads/AdMobDisplayer.kt:160]: @Climbatize noted!!

File: testapp/build.gradle
Comment (Climbatize) [reply testapp/build.gradle:146]: Thanks for the addition @munir-rakuten, much better :)

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/ads/AdMobDisplayer.kt
Comment (Climbatize) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/ads/AdMobDisplayer.kt:148]: You still have a forced unwrapping there ;)

**Review by Climbatize — DISMISSED** [review]
Looks good to me

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/ads/AdMobDisplayer.kt
Comment (munir-rakuten) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/ads/AdMobDisplayer.kt]: updated it,  thanks 🙂

**Review by khairul-alam-licon — APPROVED** [review]
Looks good 👍

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/ads/AdMobDisplayer.kt
Comment (corycaywood) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/ads/AdMobDisplayer.kt:24]: I wonder if it would make more sense to just make `AdmobDisplayerLatest` be it's own independent implementation of `MiniAppAdDisplayer`? It seems like this `AdMobDisplayer` class is becoming complicated (and you need to use reflection to check for `AdMobDisplayerLatest`).

But instead you could just implement the `MiniAppAdDisplayer` interface in `AdmobDisplayerLatest` and then if the App wants to use the latest Admob, then they should use `AdmobDisplayerLatest` instead of `AdMobDisplayer`. And then maybe we could rename `AdMobDisplayer` to `AdMob7Displayer` (and deprecate the old name using a `typealias`).

What do you think?

**Review by corycaywood — COMMENTED** [review]
It's looking good. I left a few comments below mostly about the build setup.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/ads/AdMobDisplayer.kt
Comment (munir-rakuten) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/ads/AdMobDisplayer.kt:24]: @corycaywood The problem i faced is if you implement the latest admob sdk, the old version can not be used,  if we try to use it, it will create a crash, I think its android default behaviour, if multiple version of the same dependency detected it will automatically choose the latest one, so if we give an option to choose, host app might include the `admob-latest` module but try to implement old version, it will create a crash .. thats why i implemented the way that host app has to use the latest version if it include the `admob-latest` module in the gradle file , otherwise it will use the old sdk...does it make sense?

File: testapp/build.gradle
Comment (khairul-alam-licon) [reply testapp/build.gradle:146]: `+` for `admob-latest` if its possible.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/ads/AdMobDisplayer.kt
Comment (khairul-alam-licon) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/ads/AdMobDisplayer.kt:24]: yeah, if there is a crash / conflict between two libs, probably we need to handle the exception as well.

File: miniapp/build.gradle
Comment (corycaywood) [inline miniapp/build.gradle:103]: This shouldn't be added as a dependency to `miniapp`. Instead, `miniapp` should be added as a dependency to `admob-latest`.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/ads/AdMobDisplayer.kt
Comment (corycaywood) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/ads/AdMobDisplayer.kt]: @munir-rakuten I think it's correct that the App should crash because it means the Host App implemented something incorrectly. The Host App should know to use the current `AdMobDisplayer` if they're using Ads SDK 19 and use the new one from the module if they're using Ads SDK 20. As I mentioned, we can rename `AdMobDisplayer` to `AdMob19Displayer` and deprecate the old name:
```
@Deprecated("Use AdMobDisplayer19 if you're using Google Ads SDK 19. If you're using Google Ads SDK 20 or higher, use AdMobDisplyerLatest from the admob-latest module.")
typealias AdMobDisplayer = AdMobDisplayer19

class AdMobDisplayer19(private val context: Activity) : MiniAppAdDisplayer, CoroutineScope {
```

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/ads/AdMob20Displayer.kt
Comment (corycaywood) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/ads/AdMob20Displayer.kt:28]: See my other comments: basically this class should exist in the `admob-latest` module. If the Host App needs to use Ads SDK 20+, then they should import this class from that module.

**Review by corycaywood — COMMENTED** [review]
@munir-rakuten This is looking good, but there is still something that needs to be resolved:
`AdMob20Displayer` should not be in the `miniapp` module - it should be in the `admob-latest` module. Basically I think the `miniapp` module should be unaffected by our support for Ads SDK 20. See the comments below.

File: admob-latest/src/main/java/com/rakuten/tech/mobile/admob/AdmobDisplayerLatest.kt
Comment (munir-rakuten) [reply admob-latest/src/main/java/com/rakuten/tech/mobile/admob/AdmobDisplayerLatest.kt:9]: how can i use MiniAppAdDisplayer in `admob-latest` module, there will be no link between the `miniapp sdk` and `admob-latest`

File: admob-latest/src/test/java/com/rakuten/tech/mobile/admob/RealAdmobDisplayerLatestSpec.kt
Comment (corycaywood) [inline admob-latest/src/test/java/com/rakuten/tech/mobile/admob/RealAdmobDisplayerLatestSpec.kt:30]: I think you don't need to use instrumented tests for this. Instead, you could just inject the functions which you need to mock. For example in `RealAdmobDisplayerLatest`:

```
internal class RealAdmobDisplayerLatest(
    private val context: Activity,
    private val loadInterstitialAd: (Context, String, AdRequest, InterstitialAdLoadCallback) -> Unit = InterstitialAd::load,
    // etc...
) : AdmobDisplayerLatest() {
```

File: admob-latest/src/main/java/com/rakuten/tech/mobile/admob/AdmobDisplayerLatest.kt
Comment (corycaywood) [reply admob-latest/src/main/java/com/rakuten/tech/mobile/admob/AdmobDisplayerLatest.kt:9]: That's correct - `miniapp` should have no link to `admob-latest`, but `admob-latest` should depend on `miniapp`. You need to add `implementation project(':miniapp')` to `admob-latest/build.gradle`.

File: admob-latest/src/main/java/com/rakuten/tech/mobile/admob/RealAdmobDisplayerLatest.kt
Comment (munir-rakuten) [reply admob-latest/src/main/java/com/rakuten/tech/mobile/admob/RealAdmobDisplayerLatest.kt:16]: yeah, that will do the work, i was just confused to add the whole `miniapp sdk` to use just one interface 🙂

File: admob-latest/build.gradle
Comment (khairul-alam-licon) [inline admob-latest/build.gradle:40]: seems, we don't need testImplementation of `mockito-android`, `mockito-core` & `mockito-inline` for now? 🤔

File: admob-latest/src/test/java/com/rakuten/tech/mobile/admob/AdMobDisplayer20Spec.kt
Comment (khairul-alam-licon) [inline admob-latest/src/test/java/com/rakuten/tech/mobile/admob/AdMobDisplayer20Spec.kt:21]: Do we need this? if we can remove this, we can also remove `testImplementation "org.robolectric:robolectric..."` from `build.gradle`

File: admob-latest/src/main/java/com/rakuten/tech/mobile/admob/AdMobDisplayer20.kt
Comment (khairul-alam-licon) [inline admob-latest/src/main/java/com/rakuten/tech/mobile/admob/AdMobDisplayer20.kt:60]: Can we remove `?: ""` from here?

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/MiniAppDisplayActivity.kt
Comment (khairul-alam-licon) [inline testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/MiniAppDisplayActivity.kt:176]: Maybe we can remove this line.
Also, we can tell that `AdMobDisplayer` has been deprecated in changelog, users also can use `AdMobDisplayer19` instead.

**Review by khairul-alam-licon — COMMENTED** [review]
Looks fine, I have left few suggestions, mostly for cleaning.

**Review by corycaywood — COMMENTED** [review]
Looks good! There's just one more thing that I saw - see my comment below.

File: admob-latest/src/main/java/com/rakuten/tech/mobile/admob/AdMobDisplayer20.kt
Comment (corycaywood) [inline admob-latest/src/main/java/com/rakuten/tech/mobile/admob/AdMobDisplayer20.kt:1]: I don't think this package name is correct: it should something like `com.rakuten.tech.mobile.miniapp.ads`.

Comment (munir-rakuten) [reply admob-latest/src/main/java/com/rakuten/tech/mobile/admob/AdMobDisplayer20.kt]: okay..updated it @corycaywood

**Review by corycaywood — APPROVED** [review]
Looks good 👍

**Review by khairul-alam-licon — APPROVED** [review]
Looks good 👍

**Pull request merged**
